Add hipcub::TilePrefixCallbackOp and hipcub::ScanTileState

Beanavil opened the pull request:
Added these for CUB parity as requested.

Naraenda committed:
add single pass scan operators

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,17 @@
 
 Full documentation for hipCUB is available at [https://rocm.docs.amd.com/projects/hipCUB/en/latest/](https://rocm.docs.amd.com/projects/hipCUB/en/latest/).
 
+## hipCUB-x.x.x for ROCm 6.x.x
```
```suggestion
## (Unreleased) hipCUB-x.x.x for ROCm 6.x.x
```
This ensures that it's not picked up by the release notes script until it's ready for release.

spolifroni-amd reviewed (approved):
There's a small change to the changelog that should be put in, but I'm approving anyways because the change is really small and I don't want my review to hold everything up.

ex-rzr committed:
Mark the current version as (Unreleased)

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

ex-rzr committed:
Fix building test_hipcub_single_pass_scan_operators on Windows

int8_t is not allowed as a result type of std::uniform_int_distribution.
The standard says that the effect is undefined in this case but MSVC explicitly forbids it.

umfranzw commented on hipcub/include/hipcub/backend/rocprim/agent/single_pass_scan_operators.hpp:
```diff
@@ -0,0 +1,473 @@
+    HIPCUB_FORCEINLINE HIPCUB_DEVICE
+    value_type get_prefix_forward(ScanOpT /* op */, const unsigned int tile_idx)
+    {
+        // We can ignore this because we only allow nondeterministic execution!
```
Should this comment be "we only allow *deterministic* execution"?

Snektron replied:
No; the default way of implementing lookback scan using these scan operators is fundamentally nondeterministic. A deterministic lookback scan is actually unique to HIP as far as I know - CUB does not have an equivalent mechanism. The nondeterministic variant is quite significantly faster, and for this reason, we only expose this version here.

umfranzw replied:
Ah, got it, thanks - I was confused because I'd misinterpreted an earlier comment on line 143. It makes sense now.

umfranzw commented on hipcub/include/hipcub/backend/rocprim/agent/single_pass_scan_operators.hpp:
```diff
@@ -0,0 +1,473 @@
+    value_type get_prefix_forward(ScanOpT /* op */, const unsigned int tile_idx)
+    {
+        // We can ignore this because we only allow nondeterministic execution!
+        __builtin_unreachable();
```
I'm probably just missing some context for understanding this, but I'm just wondering - if this function can be called during deterministic execution, why is the body of the function considered unconditionally unreachable? It might be nice to have a bit more of an explanation in the comments here just to make things clearer.

Snektron replied:
This body can never be reached because we force nondeterministic execution here: https://github.com/ROCm/hipCUB/blob/a590106328ef8e3aeac67ccdde0adc631228138b/hipcub/include/hipcub/backend/rocprim/agent/single_pass_scan_operators.hpp#L363
Unfortunately, we cannot simply omit this method currently, because its selected via runtime dispatch: https://github.com/ROCm/rocPRIM/blob/develop/rocprim/include/rocprim/device/detail/lookback_scan_state.hpp#L791. Note that this condition is technically compile-time known (indicated by ROCPRIM_IF_CONSTEXPR, which is not actually constexpr in C++14 mode), and so it is optimized away, but according to the C++ semantics the else-branch must still be analyzed in this case and so compilation will fail if this function is not declared. I guess an alternative would be to either base rocPRIM off C++17, or to manually rewrite this to a comptime known branch.

umfranzw replied:
Thanks for the extra context, that makes makes sense to me now - I think this is fine as-is without additional comments.

umfranzw reviewed (changes requested):
I took a look through, and just had a two quick comments/questions.

umfranzw reviewed (approved):
I think this change looks good, pending the CI tests passing.

stanleytsang-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,17 @@
 
 Full documentation for hipCUB is available at [https://rocm.docs.amd.com/projects/hipCUB/en/latest/](https://rocm.docs.amd.com/projects/hipCUB/en/latest/).
 
+## (Unreleased) hipCUB-x.x.x for ROCm 6.x.x
```
We don't know what the next ROCm release/version will be, so I'm okay with leaving these placeholders in the changelog for now, but we'll need to fix this once we know what the next release version is.

stanleytsang-amd commented:
CI tests passed except for BlockHistogram, which is most likely due to existing compiler bug.  See https://github.com/ROCm/rocPRIM/pull/672
